core/src/main/java/cz/xtf/core/openshift/HelmClients.java

```diff
@@ -0,0 +1,109 @@
+package cz.xtf.core.openshift;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Paths;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang3.SystemUtils;
+
+import cz.xtf.core.config.OpenShiftConfig;
+import cz.xtf.core.http.Https;
+
+public class HelmClients {
+
+    private static final String HELM_CLIENTS_URL = "https://mirror.openshift.com/pub/openshift-v4/clients/helm/latest";
+    private static final String HELM_FILE_NAME = "helm.tar.gz";
+
+    private static HelmBinary adminHelmBinary;
+    private static HelmBinary masterHelmBinary;
+
+    public static HelmBinary admin() {
+        if (adminHelmBinary == null) {
+            adminHelmBinary = getBinary(OpenShiftConfig.adminToken(),
+                    OpenShiftConfig.adminUsername(),
+                    OpenShiftConfig.adminPassword(),
+                    OpenShiftConfig.adminKubeconfig(),
+                    OpenShiftConfig.namespace());
+        }
+        return adminHelmBinary;
+    }
+
+    public static HelmBinary master() {
+        if (masterHelmBinary == null) {
+            masterHelmBinary = getBinary(OpenShiftConfig.masterToken(),
+                    OpenShiftConfig.masterUsername(),
+                    OpenShiftConfig.masterPassword(),
+                    OpenShiftConfig.masterKubeconfig(),
+                    OpenShiftConfig.namespace());
+        }
+        return masterHelmBinary;
+    }
+
+    private static HelmBinary getBinary(String token, String username, String password, String kubeconfig,
+            String namespace) {
+        OpenShiftBinary openShiftBinary = OpenShifts.getBinary(token, username, password, kubeconfig, namespace);
```

Comment:
`HelmBinary` is depending on `oc.config` that is being downloaded with the `oc.binary`.  The `OpenSjifts.getBinary()` calls downloads the binary, config and performs login on that config, so I found it convenient, as we'll stay in the same context as with OpenShift binary

Reply:
I guess we can have a workaround by passing env variables to helm binary. This may result in not requiring `oc.config` at all (https://helm.sh/docs/helm/helm/)

Reply:
I am not sure I follow. Maybe some chat or short call will help me. Are we trying to solve som real problem?

> I found it convenient, as we'll stay in the same context as with OpenShift binary

So is it about reusing `oc.config` path? Is solution for that just expose constant for `oc.config` and use it on both places? 

> I think we can do this with HELM_KUBETOKEN and HELM_KUBEASUSER variables

Or is it about sharing token? But we already have this parameters as input `String token, String username`

Reply:
We can 1) reuse `oc.config`, but in order to do that, we need to download it first. AFAIK only `OpenShiftBinary` uses this config and `OpenShift` is usually configured to use `OpenShiftConfig` internally (fabric8 API). `OpenShiftBinary` will download it with the binary itself, that's why I've decided to bundle `helm` binary with `oc` binary. 
2) If we don't want to re-use `oc.config` (or use it at all), we may not need to download it at all, and we can provide the environment variables to `ProcessBuilder`

AFAIK `oc.config` has `context` entries which point to user activity within namespace, so using `oc.config` can be fully replaced using 2)

Reply:
`reuse oc.config, but in order to do that, we need to download it first.` download from where? I thought when we talk about `oc.config` we talk about `~/.kube/config` local machine file filled with entries based on how `oc login` is used on that machine.

Is `oc.config` `~/.kube/config` ?

Reply:
Or in other words my understanding is we are taling here about 

```
$KUBECONFIG	set an alternative Kubernetes configuration file (default "~/.kube/config")
```
or
```
--kubeconfig string           path to the kubeconfig file
```
Described in https://helm.sh/docs/helm/helm/ . Am I missing something?

Reply:
I was talking about case of `OpenShifts.getBinary()` (https://github.com/xtf-cz/xtf/blob/master/core/src/main/java/cz/xtf/core/openshift/OpenShifts.java#L122) method, which will create empty `oc.config` in temporary directory, so the whole interaction of `OpenShitBinary` when running TS will not affect user's local `~/.kube/config`.

Reply:
Ok, so I am starting to understand it. `oc.config` is internal way of OpenShiftBinary how to keep authentication context. So you are basically delegating login logic to OpenShiftBinary. I understand that was easiest and most straighforward way.

Now I do not now if this is elegant or hack :) I tend to think it is hack :)

Conceptually I think HelmBinary and OpenShiftBinary should be independent.

I know there were cases where mystoriously in OpenShiftBinary authentication context was switching and I rather explicitely made .login() to ensure corect user is used with OpenShiftBinary. Probably it was because internal oc.config was changed.

In other words can it happen I `OpenShiftBinary.login()` with one user and I will switch user with `getHelmBinary()` call? I think that is dangerous.

But I still admit I do not see situation clear.

Reply:
@jbliznak this is discussion, why I am not sure if we want this in in this form.

Reply:
@mchoma I agree on decoupling `OpenShiftBinary` and `HelmBinary`. I tried getting rid of `oc.config` and use env variables only (7e62dadf83252aa273b763b788cd25bdf5ad025e) but looks like, in addition to `HELM_KUBEASUSER`, `HELM_KUBETOKEN` and `HELM_KUBEAPISERVER`, we need to set `HELM_KUBECAFILE` env variable as well in order to establish secure connection with the K8S server.  This brings user rather complications into the usage of `HelmBinary` IMHO.

Reply:
`oc.config` is AFAIK temporary `kubeconfig` that in our case and our method calls, will be downloaded for each instance of `HelmBinary`. `OpenShiftBinary.login()` is being called (https://github.com/xtf-cz/xtf/blob/master/core/src/main/java/cz/xtf/core/openshift/OpenShifts.java#L129) everytime the `OpenShiftBinary` would is downloaded and every call of `OpenShifts.getBinary()` downloads new `OpenShiftBinary` instance so I don't think there would be any OpenShift context interference

Reply:
We have agreed on call. It seems it is not possible fully decouple OpenShiftBinary and HelmBinary.  The reason for that is HelmBinary does not have same capabilities [1] when it comes to authenticating to OpenShift as `oc`. Helm can authenticate with 
- kubeconfig - but we need to gain it with `oc`
- token - we can use it from test.properties, but still we do not have support for trustall and successfully create connection with SSL

[1] https://helm.sh/docs/helm/helm/#options